### PR TITLE
[redis instance supports prepaid type]

### PR DESCRIPTION
**Pull request by brickzzhang**
1. redis instance supports prepaid type
2. modify wrong description "vpng" to "vpn"

commiter: brickzzhang

---------------------------------------------------

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudRedisInstance -timeout 120m
?   	github.com/terraform-providers/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/gendoc	(cached) [no tests to run]
=== RUN   TestAccTencentCloudRedisInstancesDataSource
--- PASS: TestAccTencentCloudRedisInstancesDataSource (50.89s)
=== RUN   TestAccTencentCloudRedisInstance
--- PASS: TestAccTencentCloudRedisInstance (134.96s)
PASS
ok  	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud	186.420s
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

### CODE REVIEW

File: tencentcloud/service_tencentcloud_redis.go [hunk 1]
```diff
@@ -560,6 +535,40 @@ func (me *RedisService) DestroyPostpaidInstance(ctx context.Context, redisId str
...
+		if errRet != nil {
+			log.Printf("[CRITAL]%s api[%s] fail, request body [%s], reason[%s]\n",
+				logId, request.GetAction(), request.ToJsonString(), errRet.Error())
+		}
+	}()
+	ratelimit.Check(request.GetAction())
+	// For prepaid instance, deal status synchronization will take some time so need to retry.
+	var response *redis.DestroyPrepaidInstanceResponse
+	err := resource.Retry(5*writeRetryTimeout, func() *resource.RetryError {
+		result, e := me.client.UseRedisClient().DestroyPrepaidInstance(request)
```
Comment (Himer) [inline tencentcloud/service_tencentcloud_redis.go:552]: ratelimit.Check(request.GetAction()) 放try里边

File: tencentcloud/service_tencentcloud_redis.go [hunk 2]
```diff
@@ -560,6 +535,40 @@ func (me *RedisService) DestroyPostpaidInstance(ctx context.Context, redisId str
...
+		if e != nil {
+			log.Printf("[CRITAL]%s DestroyPrepaidInstance fail, reason:%s\n", logId, e.Error())
+			return retryError(e)
+		}
+		response = result
+		return nil
+	})
+	if err == nil {
+		log.Printf("[DEBUG]%s api[%s] , request body [%s], response body[%s]\n",
```
Comment (Himer) [inline tencentcloud/service_tencentcloud_redis.go:561]: 这个log删除吧

File: tencentcloud/service_tencentcloud_redis.go [hunk 3]
```diff
@@ -571,7 +580,17 @@ func (me *RedisService) CleanUpInstance(ctx context.Context, redisId string) (ta
 		}
 	}()
 	ratelimit.Check(request.GetAction())
-	response, err := me.client.UseRedisClient().CleanUpInstance(request)
+	// Cleaning up action for prepaid instances needs to retry.
+	var response *redis.CleanUpInstanceResponse
+	err := resource.Retry(6*writeRetryTimeout, func() *resource.RetryError {
+		result, e := me.client.UseRedisClient().CleanUpInstance(request)
```
Comment (Himer) [inline tencentcloud/service_tencentcloud_redis.go:586]: ratelimit

File: tencentcloud/resource_tc_redis_instance.go [hunk 4]
```diff
@@ -606,10 +655,45 @@ func resourceTencentCloudRedisInstanceDelete(d *schema.ResourceData, meta interf
...
+		if online {
+			for k, v := range REDIS_CHARGE_TYPE_ID {
+				if v == *info.BillingMode {
+					chargeType = k
+					return nil
+				}
+			}
+			return resource.NonRetryableError(fmt.Errorf("Responsed BillingMode %d is incorrect.", *info.BillingMode))
+		}
+		return resource.RetryableError(fmt.Errorf("Deleting ERROR: Creating redis task is processing."))
```
Comment (Himer) [inline tencentcloud/resource_tc_redis_instance.go:679]: Creating redis task?

File: tencentcloud/resource_tc_redis_instance.go [hunk 5]
```diff
@@ -630,22 +714,50 @@ func resourceTencentCloudRedisInstanceDelete(d *schema.ResourceData, meta interf
...
+			log.Printf("[CRITAL]%s redis %s fail, reason:%s\n", logId, action, err.Error())
+			return err
+		}
+
+		// Deal info only support create and renew and resize, so no need to check destroy status.
+
+		if forceDelete {
+			action = "CleanUpInstance"
+			taskId, err := service.CleanUpInstance(ctx, d.Id())
+			if err != nil {
```
Comment (Himer) [inline tencentcloud/resource_tc_redis_instance.go:750]: 如果prepaid 模式下forceDelete   也需要wait DestroyPrepaidInstance任务的结果吧?

File: examples/tencentcloud-redis/main.tf [hunk 6]
```diff
@@ -26,3 +26,18 @@ data "tencentcloud_redis_instances" "redis-tags" {
...
 }
+
+resource "tencentcloud_redis_instance" "redis_instance_prepaid_test" {
+  availability_zone                   = var.availability_zone
+  type_id                             = 2
+  password                            = "test12345789"
+  mem_size                            = 8192
+  name                                = "terrform_pre_test"
+  port                                = 6379
+  charge_type                         = 1
```
Comment (gailwang) [inline examples/tencentcloud-redis/main.tf:37]: charge_type忘记改了？

File: tencentcloud/resource_tc_redis_instance.go [hunk 7]
```diff
@@ -606,10 +655,45 @@ func resourceTencentCloudRedisInstanceDelete(d *schema.ResourceData, meta interf
...
+	redisId := d.Id()
+	var chargeType string
+	errQuery := resource.Retry(20*readRetryTimeout, func() *resource.RetryError {
+		has, online, info, err := service.CheckRedisCreateOk(ctx, redisId)
+		if err != nil {
+			log.Printf("[CRITAL]%s redis querying before deleting fail, reason:%s\n", logId, err.Error())
+			return resource.NonRetryableError(err)
+		}
+		if !has {
+			return resource.NonRetryableError(fmt.Errorf("Deleting ERROR: Redis instance do not exists."))
```
Comment (gailwang) [inline tencentcloud/resource_tc_redis_instance.go:668]: delete的时候没有了不要报错，有可能是客户在控制台删除了，这时候要返回nil直接默认成功

File: tencentcloud/resource_tc_redis_instance.go (hunk 5 above)
Comment (brickzzhang) [reply tencentcloud/resource_tc_redis_instance.go:750]: DestroyPrepaidInstance返回的是dealid，但是dealid只能查create and renew and resize，之前试过

File: tencentcloud/resource_tc_redis_instance.go (hunk 4 above)
Comment (brickzzhang) [reply tencentcloud/resource_tc_redis_instance.go:679]: 这块考虑是排除那种正在创建中的状态，想在删除之前校验状态必须是online的

File: tencentcloud/service_tencentcloud_redis.go (hunk 1 above)
Comment (brickzzhang) [reply tencentcloud/service_tencentcloud_redis.go:552]: done

File: tencentcloud/service_tencentcloud_redis.go (hunk 2 above)
Comment (brickzzhang) [reply tencentcloud/service_tencentcloud_redis.go:561]: done

File: tencentcloud/service_tencentcloud_redis.go (hunk 3 above)
Comment (brickzzhang) [reply tencentcloud/service_tencentcloud_redis.go:586]: done

File: tencentcloud/resource_tc_redis_instance.go [hunk 8]
```diff
@@ -186,6 +186,27 @@ func resourceTencentCloudRedisInstance() *schema.Resource {
 				Computed:    true,
 				Description: "The time when the instance was created.",
 			},
+			// payment
+			"charge_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Default:      REDIS_CHARGE_TYPE_POSTPAID,
+				ValidateFunc: validateAllowedStringValue([]string{"PREPAID", "POSTPAID"}),
```
Comment (likexian) [inline tencentcloud/resource_tc_redis_instance.go:195]: 这里建议用常量

[]string{REDIS_CHARGE_TYPE_POSTPAID, REDIS_CHARGE_TYPE_PREPAID  }

File: tencentcloud/resource_tc_redis_instance.go (hunk 7 above)
Comment (brickzzhang) [reply tencentcloud/resource_tc_redis_instance.go:668]: done

File: tencentcloud/resource_tc_redis_instance.go (hunk 8 above)
Comment (brickzzhang) [reply tencentcloud/resource_tc_redis_instance.go:195]: done

File: tencentcloud/data_source_tc_redis_instances.go [hunk 9]
```diff
@@ -245,6 +251,8 @@ instanceLoop:
 		instanceDes["redis_replicas_num"] = instance.RedisReplicasNum
 		instanceDes["type_id"] = instance.TypeId
```
Comment (gailwang) [inline tencentcloud/data_source_tc_redis_instances.go:253]: 去掉空行

File: tencentcloud/resource_tc_redis_instance.go [hunk 10]
```diff
@@ -186,6 +186,27 @@ func resourceTencentCloudRedisInstance() *schema.Resource {
...
 				Description: "The time when the instance was created.",
 			},
+			// payment
+			"charge_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Default:      REDIS_CHARGE_TYPE_POSTPAID,
+				ValidateFunc: validateAllowedStringValue([]string{REDIS_CHARGE_TYPE_POSTPAID, REDIS_CHARGE_TYPE_PREPAID}),
+				Description:  "The charge type of instance. Valid values are `PREPAID` and `POSTPAID`.Note: TencentCloud International only supports `POSTPAID`. Caution that update operation on this field will delete old instances and create new with new charge type.",
```
Comment (gailwang) [inline tencentcloud/resource_tc_redis_instance.go:196]: `POSTPAID`.Note: 
`POSTPAID`. Note:

File: tencentcloud/resource_tc_redis_instance.go [hunk 11]
```diff
@@ -630,22 +714,50 @@ func resourceTencentCloudRedisInstanceDelete(d *schema.ResourceData, meta interf
...
+
+			return wait(action, taskId)
+		}
+
+		return nil
 	}
 
-	return wait(action, taskId)
+	log.Printf("[CRITAL] none recognized charge type %s, redis deleting action fail.", chargeType)
+	return fmt.Errorf("None recognized charge type, redis deleting action fail.")
```
Comment (gailwang) [inline tencentcloud/resource_tc_redis_instance.go:762]: 这个log打出来的信息可以合并到fmt内

File: tencentcloud/service_tencentcloud_redis.go [hunk 12]
```diff
@@ -570,8 +577,18 @@ func (me *RedisService) CleanUpInstance(ctx context.Context, redisId string) (ta
...
+		if e != nil {
+			log.Printf("[CRITAL]%s CleanUpInstance fail, reason:%s\n", logId, e.Error())
+			return retryError(e)
+		}
+		response = result
+		return nil
+	})
 	if err == nil {
 		log.Printf("[DEBUG]%s api[%s] , request body [%s], response body[%s]\n",
```
Comment (gailwang) [inline tencentcloud/service_tencentcloud_redis.go:593]: 这个log是没有干掉吗

File: website/docs/d/redis_instances.html.markdown [hunk 13]
```diff
@@ -38,6 +38,7 @@ The following arguments are supported:
 In addition to all arguments above, the following attributes are exported:
 
 * `instance_list` - A list of redis instance. Each element contains the following attributes:
+  * `charge_type` - The charge type of instance. Valid values are `POSTPAID` and `PREPAID`.
```
Comment (gailwang) [inline website/docs/d/redis_instances.html.markdown:41]: 这里要写上Default is XXX

File: website/docs/r/redis_instance.html.markdown [hunk 14]
```diff
@@ -35,8 +35,11 @@ The following arguments are supported:
...
 * `mem_size` - (Required) The memory volume of an available instance(in MB), please refer to tencentcloud_redis_zone_config.list[zone].mem_sizes.
 * `password` - (Required) Password for a Redis user, which should be 8 to 16 characters.
+* `charge_type` - (Optional, ForceNew) The charge type of instance. Valid values are `PREPAID` and `POSTPAID`.Note: TencentCloud International only supports `POSTPAID`. Caution that update operation on this field will delete old instances and create new with new charge type.
+* `force_delete` - (Optional) Indicate whether to delete Redis instance directly or not. Default is false. If set true, the instance will be deleted instead of staying recycle bin. Note: only works for PREPAID instance.
 * `name` - (Optional) Instance name.
 * `port` - (Optional, ForceNew) The port used to access a redis instance. The default value is 6379. And this value can't be changed after creation, or the Redis instance will be recreated.
+* `prepaid_period` - (Optional) The tenancy (time unit is month) of the prepaid instance, NOTE: it only works when instance_charge_type is set to `PREPAID`. Valid values are 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 24, 36.
```
Comment (oliverpei) [inline website/docs/r/redis_instance.html.markdown:42]: instance_charge_type -> `charge_type`

File: tencentcloud/resource_tc_redis_instance.go [hunk 15]
```diff
@@ -455,6 +489,12 @@ func resourceTencentCloudRedisInstanceRead(d *schema.ResourceData, meta interfac
 	}
 	_ = d.Set("tags", tags)
 
+	for k, v := range REDIS_CHARGE_TYPE_ID {
+		if v == *info.BillingMode {
+			_ = d.Set("charge_type", k)
```
Comment (Himer) [inline tencentcloud/resource_tc_redis_instance.go:494]: REDIS_CHARGE_TYPE_ID建议定义个map

var REDIS_CHARGE_TYPE_ID = map[string]int64{
	: 0:REDIS_CHARGE_TYPE_POSTPAID
:  1:	REDIS_CHARGE_TYPE_PREPAID
}

File: tencentcloud/resource_tc_redis_instance.go [hunk 16]
```diff
@@ -630,22 +704,53 @@ func resourceTencentCloudRedisInstanceDelete(d *schema.ResourceData, meta interf
...
+		if err = wait("DestroyPostpaidInstance", taskId); err != nil {
+			return err
+		}
 
-	action = "CleanUpInstance"
-	taskId, err = service.CleanUpInstance(ctx, d.Id())
-	if err != nil {
-		log.Printf("[CRITAL]%s redis %s fail, reason:%s\n", logId, action, err.Error())
-		return err
+	} else if chargeType == REDIS_CHARGE_TYPE_PREPAID {
```
Comment (Himer) [inline tencentcloud/resource_tc_redis_instance.go:719]: chargeType 没有其他类型了
-------------
"charge_type" 有 ValidateFunc: validateAllowedStringValue([]string{REDIS_CHARGE_TYPE_POSTPAID, REDIS_CHARGE_TYPE_PREPAID}),

File: tencentcloud/resource_tc_redis_instance.go [hunk 17]
```diff
@@ -630,22 +704,50 @@ func resourceTencentCloudRedisInstanceDelete(d *schema.ResourceData, meta interf
...
+			log.Printf("[CRITAL]%s redis %s fail, reason:%s\n", logId, "DestroyPrepaidInstance", err.Error())
+			return err
+		}
+
+		// Deal info only support create and renew and resize, need to check destroy status by describing api.
+		if errDestroyChecking := resource.Retry(20*readRetryTimeout, func() *resource.RetryError {
+			has, isolated, err := service.CheckRedisDestroyOk(ctx, d.Id())
+			if err != nil {
+				log.Printf("[CRITAL]%s CheckRedisDestroyOk fail, reason:%s\n", logId, err.Error())
+				return resource.NonRetryableError(err)
```
Comment (Himer) [inline tencentcloud/resource_tc_redis_instance.go:730]: resource.RetryableError

**Pull request merged**